### ISSUE TITLE
Standard library factory functions are not consistent

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#14291]
Related https://github.com/chapel-lang/chapel/issues/12712

Request for consistent identifer-naming style for standard-library factory functions.

The ones I've seen so far include: `makeRandomStream`, `newBlockDist`, and `createStringWith*`.

Alternatively/preferably, define them as type methods instead. Then everything can be e.g., `BlockDist.new(...)`

**Comment by bradcray** [comment]
@e-kayrakli: I'm particularly interested in whether we considered making the string factory functions into type methods on string (e.g., `string.createWith*`) to avoid relying on global standalone functions.  It seems obvious in retrospect, but is there something I'm missing?  (not to say that the others aren't similarly interesting, just that this was the most recently added case).

**Comment by mppf** [comment]
@BryantLam - Is the difference in terms `make` / `new` / `create` what is bothering you here? I can imagine many other ways they might not be consistent but if it is simply this term I think it's worth saying that directly in the issue description.

**Comment by mppf** [comment]
#6698 is also related.

**Comment by e-kayrakli** [comment]
@bradcray -- I don't recall any discussions about making these factory functions type methods for strings/bytes. I couldn't see anything like that in the design discussion we had over at #13818 

For those two types at least, the idea of converting factory functions to type methods seems appealing to me.

**Comment by BryantLam** [comment]
> BlockDist.new(...)

While I would prefer `new` for symmetry, `new` is a keyword and cannot be used as an identifier. `make` is my next preference only because it's shorter than `create`.

**Comment by ben-albrecht** [comment]
I am in favor of standardizing factory functions as type methods with a common verb used. Despite being slightly longer, I have a slight preference towards `create` over `make`, due to being clear and less overloaded of a term than `make`.

Some interesting points are made in answers to a [SO question](https://stackoverflow.com/questions/3368830/how-to-name-factory-like-methods) on naming factory functions:

>'Create' fits the feature better than most other words. The next best word I can think of off the top of my head is 'Construct'. In the past, 'Alloc' (allocate) might have been used in similar situations, reflecting the greater emphasis on blocks of data than objects in languages like C.
> 
>'Create' is a short, simple word that has a clear intuitive meaning. In most cases people probably just pick it as the first, most obvious word that comes to mind when they wish to create something. It's a common naming convention, and "object creation" is a common way of describing the process of... creating objects.

...

>'Build' and 'Make' are common terms for processes relating to compiling code, so have different connotations to programmers, implying a process that comprises many steps and possibly a lot of disk activity. However, the idea of a Factory "building" something is a sensible idea - especially in cases where a complex data-structure is built, or many separate pieces of information are combined in some way.

**Comment by bradcray** [comment]
I tend to agree with Ben's preference for `create` over `make` (but I also type really, really fast, so that extra character is unlikely to affect me as much as some people).

An interesting case to wrestle with is whether factory functions whose names contain more than just `create()` should also use `create` or whether they could use `newBlah` instead.  E.g., the string/bytes factories will likely continue to have additional words in their names differentiating the different modes of creating the value.

I'm not wild about this, but just to put it out there:  We could also support `type.new()` in the parser and have it desugar to some well-defined method name (similar to what we do for `.domain` on arrays today, though that is also controversial and perhaps likely to go away).  But that introduces some pretty significant cases for confusion so I'm not advocating for it so much as putting it out for people to spit on or applaud as they see fit.

**Comment by BryantLam** [comment]
I'm okay with `create()`. I type fast too and I agree that "make" is overloaded.

> We could also support `type.new()` in the parser and have it desugar to some well-defined method name

I could go either way on this. I like it, but don't like the complexity that it brings. `create` is a fine trade-off to make the compiler have less special-casing.

**Comment by mppf** [comment]
My 2 cents: `create` works for me.

I'd be about equally happy with `type.new()`. In particular I think that `new` being a keyword will make it more likely for people to use the convention (although they might also think that `type.new()` isn't just a type method). Anyway `type.new()` has the advantage of requiring fewer terms for users.

**Comment by e-kayrakli** [comment]
Just leaving a note that I prefer "create", too. All the reasons that I can think of are listed above, but just to repeat the most important ones to me:

- "create" feels the most natural for me for a factory function
- "make" would be overloading the term
- "new" doesn't sound like just a function but something more special

I started working on this and got `string.createWith*Buffer` working in a [branch](https://github.com/chapel-lang/chapel/compare/master...e-kayrakli:standardize-factory-functions), and hoping to convert others and open a PR soon.

The only tiny drawback that I can see is having two deprecated ways of creating strings on master (`string.init` and `createStringWith*`) but `string.init` should be removed completely in the next release. So, I don' think it is too ugly to worry about.

**Comment by e-kayrakli** [comment]
Another factory function that needs to change is [makeIndexBuffer](https://chapel-lang.org/docs/master/builtins/ChapelArray.html#ChapelArray.makeIndexBuffer).

This one is slightly different in the sense that it needs to be called on a sparse domain, so it shouldn't be a type method. But still the name should be changed to `createIndexBuffer` to keep things consistent.

**Comment by e-kayrakli** [comment]
After some discussions under PRs and offline with @mppf and @ben-albrecht I want to bring up one potentially complicated case with the Random module. We currently have:

- A type alias `RandomStream` that aliases the default RNG (that is PCG)
- `class PCGRandomStream`
- `class NPBRandomStream`
- `makeRandomStream` that is deprecated in favor of `createRandomStream` (via #14452)

The question here is whether `createRandomStream` (or in general `createType`) is an acceptable variation of the convention we are trying to establish here. The reasons I am (partly motivated by @mppf) reluctant to add `RandomStream.create`:

- `createRandomStream` has an `algorithm` argument that allows the programmer to choose a different algorithm than the default. That would not be possible with `RandomStream.create`.
- This interface is also fits well with other Random module convenience functions like `fillRandom`, `shuffle` etc, where they also take an `algorithm` argument
- A potential `[PCG|NPB]RandomStream.create` does not really add anything on top of `new owned [PCG|NPB]RandomStream`, as opposed to other other factory functions discussed here:
  - `string`/`bytes` ones add more verbosity/clarity
  - distributed domain/array ones avoid a lot of typing

Not as important but also changing `mySparseDomain.makeIndexBuffer` to `mySparseDomain.createIndexBuffer` is another example where a `create*` method is not a type method. And I don't see a way to make it a type method without making it look contrived. 

The bottom line is, I am personally in favor of viewing the functions/methods that start with `create` as a convention rather than making the convention strictly about type methods starting with `create*`

**Comment by BryantLam** [comment]
>     * That would not be possible with `RandomStream.create`.

I'm probably missing something but why doesn't making it a type method work?

> The bottom line is, I am personally in favor of viewing the functions/methods that start with `create` as a convention rather than making the convention strictly about type methods starting with `create*`

One concern is that if multiple conventions are applicable to a type, which would a developer implement? Does the developer prefer one of them (e.g., the type method), or do you implement all of them (e.g., and also a global function)? Will a user remember which was implemented? I'd rather not have to constantly look up in the documentation to find which are implemented (but maybe factory functions are more exceptional than normal and won't be a common idiom). If type methods are not applicable to some usage patterns, I think I'd rather have all global functions than some be global and others be type methods just so I don't have to remember what style of factory functions Chapel chose to implement.

**Comment by e-kayrakli** [comment]
> I'm probably missing something but why doesn't making it a type method work?

Choosing the RNG algorithm via a function argument would not work with type method. Instead you'd have to use:

```
RandomStream.create(); // get an instance of default RNG (PCG)
PCGRandomStream.create(); // get an instance of PCG RNG
NPBRandomStream.create(); // get an instance of NPB RNG
```
Having to call different function to get different RNG streams might make user code clunkier than just passing an argument to a unified factory function. (Although, I am not sure if an application would need more than one type of random streams. And, currently you can control which one is default by the compiler flag `-sdefaultRNG=(RNG.PCG|RNG.NPB)` and in the application code use the first one of the above all the time.)

That being said, I believe your concern is valid. If we want to restrict the factory functions to be all type or non-type methods I see the following options:

1. Forget `type.create()` and adopt `createType()` as the factory function convention.
2. Keep `type.create()` as the convention
  2a. Remove any factory functions from the Random module. Have the current utilities `fillRandom`, `shuffle` and `permutation` as-is with `algorithm` argument. But make the user use initializers to create a random stream. 
  2b. Make all the random module utility functions type methods on different RNG algorithms, instead of exposing that via an argument. This would result in some extra typing as well:

```chapel
fillRandom(myArr);  // default algorithm
// will turn into
RandomStream.fillRandom(myArr);
```

Some of these can be discussed as a separate issue, but the ideas here can be pertinent to other (future?) modules when implementing factory/utility functions. So I am keeping the discussion here for now.

**Comment by mppf** [comment]
The factor functions mentioned in the issue exist for different reasons. Of course `new MyType` is the "normal" way to create something. This issue is all about exceptions.

 * `makeRandomStream` uses an argument to decide which type to create, so isn't implementable with`new`
 * `newBlockDist` hides some implementation details that cause `new Block` to not generally do what is desired
 * `createStringWith*` include more information about *how* the string should be created in the function name

I think it is reasonable to have consistent terminology for alternatives to `new Something` (e.g. "create") but I don't think it's reasonable to expect them to always be type methods or have a specific name.

One big reason for that is that I consider only the 1st of these to be a *factory* pattern. The typical *factory function* examples that I know of actually takes a string argument specifying which thing. E.g. `createEncoder("UTF-8")` or `createHashFunction("SHA-256"). These functions do not always return the same type, but it is common for them to return a type meeting certain requirements (e.g. implement an interface or subtype of parent class).

Back to this:

> >  I'm probably missing something but why doesn't making it a type method work?

> Choosing the RNG algorithm via a function argument would not work with type method. Instead you'd have to use:

It could work with a type method; it would just be strange. We don't have a "Random Stream Factory" type. So, we could either add such a type (but... what's the point of doing so?) or we could make the RandomStream implementations able to return a different type according to a `param` argument. E.g. `proc type RandomStream(param algorithm = defaultRNG)`.

Anyway, I don't think that we should require that all factory functions are or are not type methods, because the reason for having the factory function is pretty different in these different cases.

**Comment by mppf** [comment]
> Make all the random module utility functions type methods on different RNG algorithms, instead of exposing that via an argument. 

We might decide that this is a better idea than the alternatives. If we chose this path, we could have a function returning a type and accepting the `param algorithm`, e.g. `proc getRandomStreamType(param algorithm) type`.

**Comment by ben-albrecht** [comment]
I think we've ended up in this situation due to an ambiguity in the Random module interface: `RandomStream` is both referred to as a type alias to the default RNG, but also as a common interface, as if it were a superclass to the random streams, which it is _not_. As a result, we are trying to justify an exception (`createType()`) to the exception (`Type.create()`) of creating objects with `new Type()`.

From here, we could accept the Random interface as it is and allow exceptions to the factory function standard to work around it, or we could try to address the issues in the Random interface.

As an interface-consistency zealot, I personally lean towards the latter. I think @e-kayrakli's suggestions above (specifically bullet 2) seem to be going in the right direction.

If `RandomStream` is a type alias to `PCGRandomStream`, then it doesn't make sense for `createRandomStream()` or `RandomStream.create()` to create something that is _not_ `PCGRandomStream`

> Remove any factory functions from the Random module. Have the current utilities `fillRandom`, `shuffle` and `permutation` as-is with algorithm argument..

The module-level functions `fillRandom()`, `shuffle()` should probably be reserved for the module-level random stream(s) when those get added (https://github.com/chapel-lang/chapel/issues/7225), so I think it is reasonable to require calling them as methods on existing random stream objects otherwise. 

Perhaps we should spin off a separate issue to discuss specifics of Random interface design further.

**Comment by BryantLam** [comment]
I agree with @ben-albrecht.

> * `newBlockDist` hides some implementation details that cause `new Block` to not generally do what is desired

This seems more like a problem with the default behavior of `new Block(...)` if users are generally expected to use `newBlockDist`/`BlockDist.create()` instead.

> * `createStringWith*` include more information about _how_ the string should be created in the function name

Yeah, there was a discussion about this approach versus an enum-argument variant in https://github.com/chapel-lang/chapel/issues/13818#issuecomment-523626592. 

This case is also somewhat different since it deals with FFI behaviors. Another approach could have been to make new types like [`OwnedString` or `BorrowedString`](https://github.com/chapel-lang/chapel/issues/13818#issuecomment-523606745) and then implement something akin to Rust's [`From` trait](https://doc.rust-lang.org/std/convert/trait.From.html): `OwnedString.from(c_str)`.

> E.g. `createEncoder("UTF-8")` or `createHashFunction("SHA-256")`. These functions do not always return the same type
> ...
> or we could make the RandomStream implementations able to return a different type according to a `param` argument. E.g. `proc type RandomStream(param algorithm = defaultRNG)`.

I'm okay if `RandomStream(int).create(algorithm = RNG.NPB)` gave me a `NPBRandomStream`; it is what I asked for.

> Anyway, I don't think that we should require that all factory functions are or are not type methods, because the reason for having the factory function is pretty different in these different cases.

In general, I would prefer type methods because it avoid polluting the global namespace. Maybe this helps the compiler find viable candidates faster? I don't know. It's also slightly easier for me to reason about where a type method (should be) defined.

But I also don't really care that much as long as everything is consistent. I do think that global functions would be easier for most people to use since that's what people are generally accustomed to already, but
1. to me, there's an elegance that `RandomStream(int).create()` requires that the type/param fields of the type are specified in the left side rather than in the method argument list. Maybe not everyone agrees that it's easy to learn or to use, though.
2. the "exceptional" cases noted also seem like they're just interfaces that could have been designed differently / more consistently to begin with. Having global functions just makes it more loosely defined as to what the consistent convention is, which means I usually have to go look it up.

**Comment by e-kayrakli** [comment]
I summarized the Random module specific discussion in a proposal for that module under #14589, I am holding off merging the PR that changes other factory functions (#14449) until the discussion on the Random module is settled to avoid potential back-and-forth.

**Comment by mppf** [comment]
I feel that the main objection I have to the direction this is going has not been sufficiently discussed.

> The typical *factory function* examples that I know of actually takes a string argument specifying which thing. E.g. `createEncoder("UTF-8")` or `createHashFunction("SHA-256")`. These functions do not always return the same type, but it is common for them to return a type meeting certain requirements (e.g. implement an interface or subtype of parent class).

If we insist that all factory functions are type methods, what type do they apply to? Is it the interface / parent class? Will it always be the case that we can name an interface / parent class? For example, in Chapel, it is possible to have generic functions that return a type based on a `param` even without any interface or parent class. Are we saying that the standard library may not use that pattern?

Broadly speaking, isn't a factory function any function that returns a newly constructed thing? `open` fits into that category - it returns a newly constructed open file. And yet I think it would but utterly unreasonable to insist the function be named `file.create()` instead of `open()`. (at the very least because "create" means something totally different in the context of the filesystem). For that matter doesn't `+` on integers return a newly constructed integer, so shouldn't we be writing `int.createSum(1, 2)` instead of `1 + 2` ?

I think it's reasonable to have a broad convention, like "use the term *create* in factory functions unless you have a good reason not to". However I think that going beyond is a really bad idea. (I think *any* style guidance that does not include "unless you have a good reason not to" is a bad idea).

**Comment by BryantLam** [comment]
> I think it's reasonable to have a broad convention, like "use the term _create_ in factory functions unless you have a good reason not to". However I think that going beyond is a really bad idea. (I think _any_ style guidance that does not include "unless you have a good reason not to" is a bad idea).

Agreed.

I care that a consistent convention exist; I don't think I particularly care enough about what the convention should be (though I clearly have opinions). In a way, this issue brought to light that the `Random` module was designed differently than other modules, maybe for good reasons. I don't know enough about `Random` to comment on it.

> Will it always be the case that we can name an interface / parent class?

I would hope that the answer is "yes" only because constrained generics would require it in most cases. For example, `string` and `bytes` need a common interface for use in constrained generics with a common set of string operations to interchange between these two types.

Is there an example of a set of related types you can think of that a named interface wouldn't make sense? 

> For example, in Chapel, it is possible to have generic functions that return a type based on a `param` even without any interface or parent class. Are we saying that the standard library may not use that pattern?

If there's a good reason for this to exist, it shouldn't be disallowed. I'm having a hard time thinking of a good function factory function that returns unrelated types.

**Comment by e-kayrakli** [comment]
After some offline discussions with @ben-albrecht @mppf and @bradcray I think we shouldn't rush into a decision on this issue.

From design standpoint, these are probably better handled with a wider style/conventions discussion. An example of this is under #6698. Moreover, (as particularly pointed out by @mppf), some of the standard library interface will be changed when we have constrained generics (interfaces) as the language feature.

From 2.0 standpoint, string and bytes factory functions are more part of the language than the others. But still, on the spectrum of language feature vs library feature, I'd argue that they are in the gray area and even closer to being a library feature.

**Comment by mppf** [comment]
@BryantLam -

Re

> For example, string and bytes need a common interface for use in constrained generics with a common set of string operations to interchange between these two types.

Note that we are discussing whether string and bytes return the string/bytes type or a different type when doing indexing on #13901. While either way we could have an interface covering some of the functions... an interface for string/bytes that doesn't include indexing would seem not very useful. The interface could conceivably have an associated type for the return/yielded type... Anyway if having a single interface for string/bytes is important to you that might be worth saying on #13901.

**Comment by e-kayrakli** [comment]
Type methods would be great for string/bytes programming. We have a ton of generic code between the two types, which at some points look like

```chapel
if t == string then
  createStringWithNewBuffer(...);
else
  createBytesWithNewBuffer(...);
```

It'd be great if we could just:

```chapel
t.createWithNewBuffer(...);
```

**Comment by aconsroe-hpe** [comment]
In the DateTime module review, a similar question arose for
  * `datetime.fromtimestamp`
  * `datetime.utcfromtimestamp`

Where at first the timestamp felt like it could be an overload to `proc datetime.init(timestamp:real)`, but with UTC in the mix, it raises the question whether it was better to leave as a named function like that? Or perhaps add an argument like `isUtc:bool` but that doesn't seem very appealing.

And as it sounds like with the other cases mentioned here, there is tension between making these standalone functions or `proc type` factories.

**Comment by bradcray** [comment]
Now that we've been living in the `createFoo()` world for awhile, I've been noticing myself wince each time I use such calls because I find that I want to write `newFoo()` instead since it's shorter, fewer syllables when read out loud, and more symmetric to `new C()` or `new R()` as a means of creating a new thing.  I.e., for me it raises questions of "What is the technical distinction between the uses of the terms `new...` and `create...` in the Chapel language?", where the answer essentially feels like "none."

That said, we took a straw poll in a deep-dive today, and I was the only person on the team who seemed to feel this way. This is good because it means that we don't need to change things that use `create` today.  But I still wanted to lodge this observation here as a means of putting it out of my head and to rest.

**Comment by lydia-duncan** [comment]
We took a look at all our factory functions from a high level perspective today, to make sure we were still on board with decisions we had made and move ones that didn't have a decision yet further forward.

Here's the summary of the status on main:
<details>
Standalone, start with create:
- createRandomStream (Random module)
- createString/BytesWithBuffer, etc (String and Bytes modules)
- createVersion (Version module)

Not for 2.0, but not using create yet (should they be adjusted for 2.0?):
- newBlockDom (BlockDist module)
- newBlockArr (BlockDist module)

Methods (usually type methods except for one), not using create:
- domain.makeIndexBuffer (ChapelDomain module, Domain section of spec, instance method, not for 2.0)
- SystemError.fromSyserr (OS module)
- datetime.fromtimestamp (DateTime module)
- datetime.utcfromtimestamp (DateTime module) and a few others like them

standalone but not using create:
- open/openfd/openfp/opentmp/ etc (IO module)
- spawn (Subprocess module)
- spawnshell (Subprocess module)
</details>

Looking at this at the big picture level, we've decided
- We like open/spawn generally and are okay with keeping them in that strategy
- We like create as the prefix
- We are okay with having both standalone and type methods as appropriate, but tend to use standalone in most cases
- domain.makeIndexBuffer should be renamed to domain.createIndexBuffer even though it is not for 2.0, to avoid confusion
- SystemError.fromSyserr should be made a standalone function and renamed to createSystemError
- We still need more discussion on the datetime type methods

I will also update our style guide to include an entry with this recommendation

**Comment by dlongnecke-cray** [comment]
I would like to advocate strongly in favor of using type methods over standalone factory functions within our modules whenever possible/sensible.

I think it makes great sense for a factory function which produces a type to live within the namespace of the type. It shields the factory and means that users only have to worry about conflicts on the type name.

I don't really buy the argument of

```chapel
var x = Module.createFooForThing();
```

Being more elegant than

```chapel
var x = Module.Foo.createForThing();
```

It seems like the difference is one additional `.`, but in return you get all the safety that the type namespace provides and don't have to worry about name conflicts for individual function names.

Additionally, it seems like a step backwards for a module containing multiple types, each with their own factory functions, to have all those factories living at the module level:

```chapel
record foo { ... }
record bar { ... }

proc createFooForThing() { ... }
proc createBarForThing() { ... }
```

Adding additional characters to the symbol name seems like a very C-like thing which is required because C doesn't have namespaces. But Chapel has a rich namespace support, and what is the point of type methods if not to mimic the "class namespace protected factory methods" in C++? It seems like a great use-case for them.

I'm not sure if the question of "do consumers have to write `Module.Foo.createForThing`" is particularly relevant, as it seems like the only portion of the expression that matters in this discussion is `Foo.createForThing` vs `createFooForThing`. How the user chooses to shorten the prefix is up to them - I would hope that library writers have the freedom to make liberal use of `use` within implementation modules as is done with `using` in C++.

**Comment by bradcray** [comment]
> I don't really buy the argument of
> 
> ```
> var x = Module.createFooForThing();
> ```
> Being more elegant than
> ```
> var x = Module.Foo.createForThing();
> ```

I'm not going to say "more elegant" so much as "more succing and flexible across coding styles."  Specifically, if I'm a fast-and-loose, trying-to-sketch-scripty-code quickly type of user, using a standalone function permits me to either:
* lean on `use Module;` and just write `createFooForThing();` or 
* be a principled, fully-qualified, safety-conscious type of user, lean on `import` and write either:
  * `import Module.createFooForThing;  createFooForThing();`
  * `import Module;  Module.createFooForThing();`

In contrast, if we use a type method, both user profiles are required to write it in a less succinct way and I'm not sure I'm convinced that the likelihood of conflicts is so great as to really be a problem, especially in an import-based world (I'm discounting `use` since it's not as good w.r.t. conflicts anyway).  That is, there's no way to simply write `createFooForThing();` once it's a type method.  While I buy the "it's better organized" argument, I'm not sure I'm willing to throw out succinctness for it.

I'd like to take this discussion from abstract `Module`, `Foo`, `createForThing()` calls to a specific example, though, because I generally don't like factories anyway, relative to initializers and casts.  Do you have an existing standalone factory routine in the standard modules that you'd like to use as a basis for this discussion so we can talk about how it would look in different forms and reason about what the opportunities for conflicts are?

**Comment by dlongnecke-cray** [comment]
I think that factories are important/elegant in the cases when the initialization is complex and requires state that needs to be computed prior to calling the initializer.

A great example I can point to you off the top of my head (granted, it's not Chapel code, but I think this is a more universal pattern): in `dyno` the constructor for a UAST node takes a ton of magic numbers which are essentially just indices into a children array. Those indices are computed by the factory, and it's a case where it's really not feasible for the user to call the constructor because computing the indices is tedious and fragile.

A knock against me - in C++ one reason we use factories is because we have to worry about inline field initialization, but not so in Chapel (you can compute stuff while assigning to fields in Chapel, right? As long as it's not using a field before `this.complete()`).

I think that factories are useful because they document the context of the construction as well as offer a means of separation of concerns for the type implementor. A good example I can think of is we offer a different way of constructing a parser in dyno: `Parser::createFromFile(string)` vs `Parser::createFromString(string)`. (Another relevant example are all the `string.create` factories.)

However I'm not against calling `new` and I think that in the common/simple case it should be the preferred way to construct a value. But when initialization of a type starts to get super complex, I think it's good to have the factory pattern in your back pocket if you need it.

---

I'd be fine using the `DateTime` factories we talked about offline as an example, I also thought that `string` ought to be `string.createWithBorrowedBuffer()`. (The other is `DateTime.date.fromOrdinal()` I believe.)

W.r.t. to the `import` examples, I'm not seeing how

- A1) `import DateTime; DateTime.date.createFromOrdinal();`
- A2) `import DateTime; DateTime.createDateFromOrdinal();`
- B1) `import  DateTime.date; date.createFromOrdinal()`
- B2) `import DateTime.createDateFromOrdinal(); createDateFromOrdinal()`

Are terribly different semantically other than the `import <type>` brings all the methods on the type into scope, but I think we already have that problem in the standalone world because we need visibility of `Foo` for the factory to produce a usable result (?). If we say that A2/B2 bring vision of the function but not the type (forgive me, I can't remember where we stand on this), then I am not sure if there is a way to achieve a similar thing with the type method, maybe `importDateTime.date.createFromOrdinal()`? Which still seems equivalent in the amount of typing.

It still just seems like it comes down to the extra `.` to me.

**Comment by mppf** [comment]
I also lean towards the type methods for factory functions where possible.

@bradcray I am curious if, say we had a type method `date.createFromOrdinal()`, would you be any happier with a type method if you could do `import DateTime.date.createFromOrdinal as createDateFromOrdinal` ?

While I do acknowledge that `date.createFromOrdinal()` is syntactically more complex (since it involves the `.` operator) — I don't view that as particularly challenging for people to work with. So that gets me to agreeing with @dlongnecke-cray's thought:

> It still just seems like it comes down to the extra `.` to me.

In other words, `date.createFromOrdinal()` vs `createDateFromOrdinal()`. These seem pretty much the same thing to me in terms of length / how annoying they are to type.

**Comment by bradcray** [comment]
I should start by saying that I'm not dead-set against type method-based approaches, just that I thought that's where the conversation was leaning the last time we had it, so I was arguing for it more to prevent churn than anything.  That said, I do tend to prefer standalone routines in most cases, though I'd probably use simpler naming conventions for them like `ordinalToDate()` which I prefer better than `createDateFromOrdinal()` or `date.createFromOrdinal()`.  Of course, that form doesn't work as well if the routine takes multiple arguments.

But also, remind me, why would we express this specific routine in any of these ways rather than as `new date(ordinal=...)`?  Are we just worried that people wouldn't use named arguments and the meaning would be unclear?  Or that we have multiple overloads with the same signature such that they'd be ambiguous if named arguments weren't used?

I agree that the ability to `import` a type method into a current scope would definitely soften the difference between the two approaches, though the type method names probably wouldn't be designed to be as clear in a standalone setting—i.e., `ordinalToDate()` seems succinct and clear whereas `createFromOrdinal()` is only succinct, not particularly clear.  So I wouldn't add this feature for my sake / to help with this specific case.

**Comment by bradcray** [comment]
@lydia-duncan : With Any finishing #21567  I came back here wondering whether we could close this, or if not, whether we could spawn any remaining cases off into their own issues and close it.  I think you're keeping score better than I am—what do you think remains / should be captured in its own issue?  Thanks.

**Comment by lydia-duncan** [comment]
I don't think there are any remaining cases that I'm aware of, thanks for the reminder!

**Comment by bradcray** [comment]
Thanks Lydia!  For reference, the cases I was least certain about the status of from your list above (and in too much of a rush to check) were:

> * domain.makeIndexBuffer should be renamed to domain.createIndexBuffer even though it is not for 2.0, to avoid confusion
> * SystemError.fromSyserr should be made a standalone function and renamed to createSystemError
> * We still need more discussion on the datetime type methods

**Comment by daviditen** [comment]
We reached consensus a long time ago that they should be camel-cased. The all lowercase names were deprecated in favor of `fromTimestamp` and `utcFromTimestamp` and the deprecated versions have since been removed. I'm not sure if our previous consensus needs to be changed again for this discussion. I'd hope not.

**Comment by lydia-duncan** [comment]
It's not the camel-casing I'm worried about.  When we discussed those functions, did we talk about how they relate to other factory functions and their naming from that perspective?  The comment from July 2022 indicated that we still needed to talk about them but I don't know when that later discussion happened.  If we did talk about them later but forgot to mention that their names weren't in keeping with other factory functions and talk about it from that perspective, that does unfortunately mean we need to revisit the discussion.
